src/cljs/kixi/hecuba/tabs/hierarchy/tapestry.cljs

```diff
@@ -0,0 +1,385 @@
+            [goog.string :as gstring]
+            [goog.string.format]
+            [bardo.interpolate :refer [interpolate pipeline]]
+            [kixi.hecuba.widgets.datetimepicker-small :as dtp]
+            [kixi.hecuba.tabs.hierarchy.raw-data :as raw]))
+
+
+(def x-readings 14)
+(def y-readings 24)
+(def chart-width (atom 800))
+(def chart-height (atom 600))
+(def chart-data-chan (atom nil))
+(def chart-fill-oor-cells? (atom true))
+(def default-gradations 20)
+(def invalid-cell-colour "#555555")
+
+(def colour-scheme
+  "http://colorbrewer2.org/?type=diverging&scheme=RdYlBu&n=10"
+  [[49 54 149] [69 117 180] [116 173 209] [171 217 233] [224 243 248] [254 224 144] [253 174 97] [244 109 67] [215 48 39] [165 0 38]])
+
+(defn generate-palette
+  "Generates n colours based on the gradations and color range specified - EXPENSIVE"
+  [grads colours]
+  (let [times (map #(/ (+ % 1) grads) (range grads))
+        rgb-funcs {:red first :green second :blue last}]
+    (mapv (fn [t] (mapv (fn [[k v]] ((pipeline (map #(/ (v %) 255) colours)) t) ) rgb-funcs)) times)))
+
+(defn svg-heatmap-with-title
+  "Custom heatmap function, as we add titles to the polygons"
+  [{:keys [x-axis y-axis project]}
+   {:keys [matrix value-domain clamp palette palette-scale attribs shape]
+    :or {value-domain [0.0 1.0]
+         palette-scale viz/linear-scale
+         shape #(conj (svg/polygon [%1 %2 %3 %4] {:fill %5}) [:title %6])}
+    :as d-spec}]
+  (let [scale-x   (:scale x-axis)
+        scale-y   (:scale y-axis)
+        pmax      (dec (count palette))
+        scale-v   (palette-scale value-domain [0 pmax])]
+    (svg/group
+     attribs
+     (for [p (nd/position-seq matrix)
+           :let [[y x] p
+                 v (nd/get-at matrix y x)]]
+       (shape
+        (project [(scale-x x) (scale-y y)])
+        (project [(scale-x (inc x)) (scale-y y)])
+        (project [(scale-x (inc x)) (scale-y (inc y))])
+        (project [(scale-x x) (scale-y (inc y))])
+        (if (or clamp (m/in-range? value-domain v))
+          (palette (m/clamp (int (scale-v v)) 0 pmax))
+          invalid-cell-colour)
+        v)))))
+
+(defn heatmap-spec
+  [id heatmap-data size-x size-y lcb ucb grads]
+  (let [matrix (viz/matrix-2d size-x size-y heatmap-data)]
+    {:matrix        matrix
+     :value-domain  [lcb ucb]
+     :palette       (generate-palette grads colour-scheme)
+     :palette-scale viz/linear-scale
+     :layout        svg-heatmap-with-title
+     }))
+
+(defn int-to-dow
+  [num]
+  (nth ["S" "M" "T" "W" "T" "F" "S"] (mod num 7)))
+
+(defn int-to-tod
+  [num]
+  (let [hrs (mod num 24)
+        mins 0]
+    (gstring/format "%02d:%02d" hrs mins)))
+
+(defn set-new-heatmap-data!
+  [cursor data lcb ucb gradations dates]
+  (let [lcb (if (nil? lcb) (.floor js/Math (apply min data)) lcb)
+        ucb (if (nil? ucb) (.ceil js/Math (apply max data)) ucb)
+        midcb ((interpolate lcb ucb) 0.5)
+        gradations (if (nil? gradations) default-gradations gradations)
```

Comment:
it's not idiomatic to use `nil?` like this if an if predicate. see earlier comment.